### ISSUE TITLE
r10k hangs when running via mcollective / web hook

### ISSUE THREAD

**Issue by esalberg** [puppetlabs/r10k#224]
Environment: gitlab using the provided web hook to kick off mco to deploy environments on push and merge.  
r10k version: 1.3.4
Puppet Enterprise version: 3.3.2 (with pe-ruby, etc.)

I see the following processes running:
peadmin  26767     1  0 13:14 ?        00:00:00 r10k  
root     26790  4987  0 13:14 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment feature_branch -p
root     26827 26790  0 13:14 ?        00:00:00 git clone --reference /var/cache/r10k/http---<server>-puppet-puppet_control.git http://<server>/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/feature_branch

The feature_branch directory is created, but r10k never finishes building the new environment (per checking to see if the "modules" directory exists within environments/feature_branch).  We have hiera data in a separate environment (environments_hiera), and the hiera environment gets setup fine.

If I kill the processes and run "r10k deploy environment -pv", then everything works fine, with a bit of delay while cloning the new feature branch.  Per debug2, the delay is after:

Execute: git clone --reference /var/cache/r10k/http---<server>-puppet-puppet_control.git http://<server>/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/facts_update

This behavior may have first started when the gitlab server had too little memory to clone environments over http, exacerbated by my changing the r10k.yaml remotes (so it tried to rebuild the caches for puppet_control and puppet_hieradata).  However, the gitlab server now should have enough memory.  In the process of troubleshooting, I did end up deleting the caches and long-standing environments (all of which are back now).

Any idea what I should check / cleanup / do?

**Comment by adrienthebo** [comment]
Does it hang if you run r10k by hand?

**Comment by esalberg** [comment]
It does not hang when running by hand, which was the weird thing (see: If I kill and run r10 deploy...).  There is a delay for the clone.  It also does not hang if I do a straight git clone of puppet_control over http.

Could there be issues with protected git branches?  I unprotected the test branch and things seem to be working again.  However, that easily could be a coincidence, since deploying by hand worked, but I wasn't sure if the web hook / mco run might be more sensitive.

**Comment by adrienthebo** [comment]
What do you mean by protected git branches?

**Comment by esalberg** [comment]
A protected git branch (in Gitlab) limits which users can push to a branch (only masters and owners - to enforce good merging practices).  I'm pretty sure it doesn't mean anything else (but not 100%).

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/permissions/permissions.md

I'll be doing more development tomorrow, so I can see if the improvement is consistent.  I wouldn't really think it would be related, since r10k is cloning, not pushing, but maybe it introduces more lag or something...

**Comment by esalberg** [comment]
Sadly, I still had problems today.  The 'r10k deploy environment -pv' always works (eventually) when I call it by hand, but when called from within a script, the process just hangs.

Is there any way to set a timeout, so at least eventually the processes would die instead of hanging around?  I've found old processes just keep hanging around - for example, these are almost 3 hours old (ref: time for the grep command).  This mostly (only?) happens when cloning a new environment.  (If I kill the processes and run deploy environment by hand, it does work.)  I know that http clone for gitlab is a bit slow, but I'd like the processes to end even if they don't finish cloning.

 18:14 pts/1    00:00:00 grep r10k
 15:38 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment avamar_module -p
15:38 ?        00:00:00 git clone --reference /var/cache/r10k/http---server.com-puppet-puppet_control.git http://server.com/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/avamar_module

**Comment by adrienthebo** [comment]
The webhook and mco agent are both part of the zack/r10k module, so there are no changes that I can directly make here to affect this. If you strace the hung process, do you see any activity?

**Comment by esalberg** [comment]
Here's what the process list looks like for my secondary master where I haven't cleaned up the hanging processes by hand yet (some of the environments it's trying to deploy are finished and removed by this point):

root     11076     1  0 Nov04 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment avamar_module -p
root     11110 11076  0 Nov04 ?        00:00:00 git clone --reference /var/cache/r10k/http---server.com-puppet-puppet_control.git http://server.com/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/avamar_module
root     19828     1  0 Nov03 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment motd -p
root     19858 19828  0 Nov03 ?        00:00:00 git clone --reference /var/cache/r10k/http---linvm950-1.putnaminv.com-puppet-puppet_control.git http://server.com/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/motd
root     26531     1  0 Nov04 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment lvm_mounts -p
root     26567 26531  0 Nov04 ?        00:00:00 git clone --reference /var/cache/r10k/http---server-puppet-puppet_control.git http://server.com/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/lvm_mounts

---

From strace, all of the r10k deploy environment calls are sitting in wait, and all of the git clone commands are stuck in read.

strace -p 11076
Process 11076 attached - interrupt to quit
wait4(11110,

strace -p 11110
Process 11110 attached - interrupt to quit
read(3,

So I don't know what's going on with the git clone commands, besides there being some delay over http to gitlab for the r10k deploy environment commands that succeed.  But it would be really helpful for the r10k command to kill itself and its child git clone command off after a certain timeframe so that I don't have zombie processes hanging around forever.

I also get the feeling that using ^C to cancel a run that seems to be hanging and then running it again has a better chance of succeeding because r10k picks up where it left off in building the cache (and this issue only occurs with a new branch or renaming the git source - aka a large amount of cache building).  However, I will note that ^C leaves processes hanging as well that need to be killed.

**Comment by adrienthebo** [comment]
Could you try running the git command, piping stderr and stdout to different files, and then seeing how big they may be? There could be some quirky behavior with the pipes that r10k constructs to collect output from executed commands.

**Comment by adrienthebo** [comment]
Specifically for the commands that are hanging.  It would be ideal to test this on a fresh clone.

**Comment by esalberg** [comment]
This is a little tricky to do without disabling the webhook (which I will do - I just haven't had a chance yet).  I went on one of the masters after pushing a new branch and tried killing the git clone and "/opt/puppet/bin/ruby /usr/bin/r10k deploy environment new_env -p" processes  removing anything that had been added to the new environment , and restarting the commands by hand with output logs, but everything was successful / nothing was logged.

Just out of curiosity, are these two commands equivalent?
/opt/puppet/bin/ruby /usr/bin/r10k deploy environment -p
vs 
/usr/bin/r10k deploy environment -p

**Comment by adrienthebo** [comment]
I don't fully understand what you did - were you killing the git and r10k processes to see what happened? And how were you observing success/failure/logging? If you're wrapping these commands in the webhooks then you're both debugging the webhook and r10k which makes it much harder to pin down what's happening.

> Just out of curiosity, are these two commands equivalent?
> /opt/puppet/bin/ruby /usr/bin/r10k deploy environment -p
> vs 
> /usr/bin/r10k deploy environment -p

Mostly, it depends on if you're using different versions of Ruby; any differences in behavior would be due to differences in the runtime.

**Comment by esalberg** [comment]
The git and r10k processes never die and keep the deploy from finishing, so I have to kill them in order to do work (I've actually written a cron script to kill r10k related processes older than 15 minutes at this point).

I observe success / failure by checking to see if a modules subdirectory was created for the new environment and/or by whether the processes are too old.  On a good run, the processes never last more than a couple of minutes.

Most of the issues seem to come with the webhook when a new branch has been created - the webhook seems to work fine for updates to existing branches.  I'm also kicking off the webhook for git pushes for two repos (I'm using r10k to deploy both modules environments and hieradata environments), so maybe it's possible that multiple r10k runs are being called at the same time or something...

I'll try to disable the webhook today and run r10k deploy environment directly / get more logging / etc. if I can.

**Comment by esalberg** [comment]
I dug a bit further into what the webhook is doing (my apologies for not checking into that sooner), and its setup to call r10k via mcollective with the following command:

```
/opt/puppet/bin/mco r10k deploy #{branch} >> #{$config['mco_logfile']} 2>&1 &
```

so for my test:

```
/opt/puppet/bin/mco r10k deploy newtest >> /var/log/webhook/mco_output.log 2>&1 &
```

I disabled the webhook and then created a new branch and ran the mco command by hand on one of my test puppet masters / mco brokers.  The result is the same - processes created that sit in read state forever (except for the r10k deploy command, which is waiting on the reads).  This is not completely consistent, as the command succeeded on one of the three masters

I also noticed that there are two git fetch-pack commands that are spawned per deploy (just lucky that I named a branch with r10k today).  There are a ton of these still existing across the masters.

If I delete /etc/puppetlabs/puppet/environments/newtest and run the r10k command (seen in 'ps' when mco runs) by hand on the server, it completes successfully - environment deployed, with no processes left behind.

Works:

```
/opt/puppet/bin/ruby /usr/bin/r10k deploy environment newtest -p >/users01/me/r10k_newtest_2.out 2>/users01/me/r10k_newtest_2.err
```

 -- I tried sending output and error to different files, but I may be doing something wrong, since nothing was written to either of those files (unless I use -v, in which case info and notices go to stderr).

If I re-run the command - either from mco or directly on the server - after killing the hanging processes but without deleting the environments/newbranch directory, it no longer hangs on the initial git clone (because the initial git directory was created during the previous failed run?).

Also, I'm pretty sure that the webhook / mco command works fine as long as all the branches are pre-existing (cloned into r10k etc.).  This issue is for a new environment branch (which unfortunately we create a lot under our workflow).

I could see this possibly being an issue with the r10k mco plug-in, but it's interesting the the processes on the masters look the same whether run by mco or directly.

Processes from of mco r10k command:

```
root      6161  9400  0 14:51 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment newtest -p
root      6191  6161  0 14:51 ?        00:00:00 git clone --reference /var/cache/r10k/http---git.server.com-puppet-puppet_control.git http://git.server.com/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/newtest
root      6199  6195  0 14:51 ?        00:00:00 git fetch-pack --stateless-rpc --lock-pack --thin --no-progress http://git.server.com/puppet/puppet_control.git  efs/heads/avamar_module  efs/heads/future  efs/heads/newtest  efs/heads/r10k_otf  efs/heads/rel20141111
root      6211  6199  0 14:51 ?        00:00:00 git fetch-pack --stateless-rpc --lock-pack --thin --no-progress http://git.server.com/puppet/puppet_control.git  efs/heads/avamar_module  efs/heads/future  efs/heads/newtest  efs/heads/r10k_otf  efs/heads/rel20141111
```

processes from running directly:

```
root     25964 30760  2 15:37 pts/0    00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment newtest -p
root     25992 25964  0 15:37 ?        00:00:00 git clone --reference /var/cache/r10k/http---git.server.com-puppet-puppet_control.git http://git.server.com/puppet/puppet_control.git /etc/puppetlabs/puppet/environments/newtest
root     26042 25999  0 15:37 ?        00:00:00 git fetch-pack --stateless-rpc --lock-pack --thin --no-progress http://git.server.com/puppet/puppet_control.git  efs/heads/avamar_module  efs/heads/future  efs/heads/newtest  efs/heads/r10k_otf  efs/heads/rel20141111
root     26080 26042  0 15:38 ?        00:00:00 git fetch-pack --stateless-rpc --lock-pack --thin --no-progress http://git.server.com/puppet/puppet_control.git  efs/heads/avamar_module  efs/heads/future  efs/heads/newtest  efs/heads/r10k_otf  efs/heads/rel20141111
```

**Comment by esalberg** [comment]
fwiw, I ran the git clone command directly logging out and error separately, and all I got was output of:

Initialized empty Git repository in /etc/puppetlabs/puppet/environments/newtest/.git/

I'm not sure how to get logging to work for the mco started commands.

Also, I wanted to note that we're using two sources (one for module environments, one for Hiera environments), and I did have the webhook configured for both of those repos.  I don't know if running the deploys too close together could be an issue (likely not the root cause).

Thank you again for looking at this with me.  I really appreciate the assistance!

ETA: Adding a new environments_hiera branch does not run into an issue, as long as the environments branch (new environment) has already been deployed.  However, environments_hiera does not have a Puppetfile / modules / etc. even though it has about 300 yaml data files, so it's much simpler / smaller than the environments source.

**Comment by adrienthebo** [comment]
So if I understand you correctly, running `r10k deploy environment` works while running the mco command fails?

**Comment by esalberg** [comment]
The mco command works as long as it's not creating a new environment from a new branch of puppet_control.  It fails most of the time when creating a new environment for puppet_control.

The mco command also worked when I killed the processes and re-ran from mco without removing the (empty) /etc/puppetlabs/puppet/environments/<newenv> directory.  However, sometimes those directories get a cache corruption error after having the processes killed, in which case I have to delete the directory and start over.

Do you think it would be better for me to take these issues over to the r10k puppet module people?  I apologize if we're trying to track something down that doesn't seem to be r10k-specific itself.  I just wanted to see if there was a better way to deal with these hanging processes rather than killing them myself.

**Comment by adrienthebo** [comment]
Sorry for the delayed response; if you can't reproduce this issue by running r10k directly then I don't think this issue can be in r10k itself. Getting in touch with the folks behind the r10k module would be your best course of action. I'm closing out this issue; if it turns out that it is an issue in r10k please comment on this issue and I'll reopen it.

**Comment by esalberg** [comment]
No problem - I had been meaning to update this ticket as well.  Things are much improved with PE 3.7.2 and the latest versions of r10k / webhook / etc.  I think the biggest improvement was switching from using http to access our gitlab repos - we now use the gms module for deploy keys to contact gitlab over ssh.  R10k is much more functional for us and I have no objection to this issue being closed.

I note that I have still seen hanging git processes for days / weeks from runs that did not complete (just much less frequently) - I may keep my git process kill script around.  However, it is unclear whether that's something r10k could handle or not - I don't know if there is a way that r10k could tell its child git processes to end after a certain amount of  time (even something large like an hour), or if that is also something that should be redirected to the r10k module owners.

**Comment by adrienthebo** [comment]
I wouldn't be averse to adding an optional timeout for Git commands; sometimes things hang. On the other hand having to kill hanging Git process seems to me that there's something else going on. Killing processes would mask the problem but sometimes that's the best route.

If hanging processes continues to be a problem, feel free to file a new ticket for that - please note that the issue tracker is moving to https://tickets.puppetlabs.com/browse/RK so if you do then the new tracker would be the best place for a new issue. Thanks!
